lib/ca_certs_stubs.c

```diff
@@ -0,0 +1,49 @@
+#include "caml/alloc.h"
+#include "caml/callback.h"
+#include "caml/fail.h"
+#include "caml/memory.h"
+
+#ifdef _WIN32
+
+#include <windows.h>
+
+value ca_certs_iter_on_anchors(value v_f)
+{
+    CAMLparam1(v_f);
+    CAMLlocal1(v_encoded_cert);
+
+    HCERTSTORE hCertStore = CertOpenSystemStore(0, "ROOT");
+    if (!hCertStore)
+    {
+        caml_failwith("ca_certs_iter_on_anchors: CertOpenSystemStore returned NULL");
+    }
+
+    PCCERT_CONTEXT pCertContext = NULL;
+    while ((pCertContext = CertEnumCertificatesInStore(hCertStore, pCertContext)) != NULL)
+    {
+        if (!(pCertContext->dwCertEncodingType & X509_ASN_ENCODING))
+        {
+            caml_failwith("ca_certs_iter_on_anchors: certificate does not have expected encoding");
```

Comment:
Should this be a warning? I can imagine (but have no experience with this on Windows) a trust anchor store with "hundreds" of normal (X509_ASN_ENCODING) certificates and a single (manually inserted) differently encoded one. I'd be fine with a warning about that single differently encoded certificate, but all others are collected normally. This would be similar to the `authenticator ()` function which outputs a log message (on the warning level) if the pem decoding fails.

Reply:
This is defensive code since this is the only possible encoding for certificates is `X509_ASN_ENCODING` at the moment.

In all of the code samples I've seen to export the list of trust anchors, it's actually assumed that the internal buffer is DER-encoded (see for example https://golang.org/src/crypto/x509/root_windows.go near line 314). So we could even remove that.

Maybe we can refine the failure mode if this becomes a problem?

Reply:
Sure, thanks for your comment. I was not aware that the only defined encoding is X509_ASN_ENCODING.